### DISCUSSION TITLE
Read-only token feedback

### DISCUSSION THREAD

**Discussion by llibdude** [tinacms/tinacms#2671] [Experimental]
Super excited for read-only tokens. I have created this issue to capture my feedback/questions/etc.

- **Experimental Autogenerated GrqphQL Client Support**
I am using the experimental autogenerated graphql client
  ```
   const client = ExperimentalGetTinaClient()
   const campaignsListData = await client.getCampaignsList()
  ```
  In my opionion, i think it might be nice if I could opt into leveraging the read-only token.  For `staticRequest` types the code above is fine, but when I want to leverage read-only tokens for client side use of `ISR` maybe I could add a flag param that indicates to use the token. Something like:
  ```
    const client = ExperimentalGetTinaClient()
    const campaignsListData = await client.getCampaignsList(true)
  ```  
  Instead of `true`, the token could be passed, but I'm thinking if the token is stored in an env variable it can be used from inside the client.
- **Wildcard git branches security support - high priority**
I don't think i have a use case for only allowing specific git branches to use a read-only token. For me, when someone authors a page they will create a new branch for that page to dev on before merging back to main.  I wouldn't want to worry about managing access in this case.  This one is pretty high up on my personal priority list.

  Wild card support, in my case just `*`, would be a must have for me.
- **Bug with token branch whitelist**
It would seem from some investigation that if I am trying to use a branch that is not the first one in the whitelist, the requests fail with a `no access to branch` error.  In my example i am trying to use the branch `tina-campaigns`.  See image below (tokens have since been cycled)
![image](https://user-images.githubusercontent.com/5255924/157531284-19aee7a8-84dc-4649-b53c-b4e17684d131.png)
First token fails, second token works.
Actually after talking with logana on discord its likely a trailing whitespace issue in the branch name caused by me typing an extra space 
- **This fixes an issue with 404 errors after creating new pages**
This is a great use case to allow the admin page to directly navigate to a freshly created page, see https://github.com/tinacms/tinacms/discussions/2630#discussioncomment-2311923 for context.  I would love to be able to auto navigate to newly created pages, even if its some kind of opt-in feature since not everyone may use read-only tokens.
- **Docs typo-ish improvement**
When doing the `data = await fetch`, the docs example is missing doing a `await data.json()`. A minor thing but caught me for a few minutes as i was brain dead copy pasting things into my codebase.
![image](https://user-images.githubusercontent.com/5255924/157550740-3a8e51e5-7680-4a00-9f07-20aec81aae57.png)

That's it for now. I do have it working, definitely the wildcard branch name thing is high on my priority list to be able to fully use this feature.  Otherwise awesome job!

**Comment by llibdude** [comment]
I'm noticing even if I just save a document after some changes using the tina editor, and refresh the page, it 404's until the commit goes through CI when not using the read-only token.  Maybe I wouldn't expect the changes to be live, but on refresh couldn't I still see the old document since the document still exists on file system? (this is with experimental data fwiw).

Either way I would love love love wildcards. I have created a specific issue for this feature https://github.com/tinacms/tinacms/issues/2794

**Reply by jamespohalloran** [reply]
This is all really great feedback! Thanks so much for this.

> Experimental Autogenerated GrqphQL Client Support

I think this should be improved with [this ticket](https://github.com/tinacms/tinacms/issues/2778), is that right @logan-anderson ?

> Wildcard git branches security support - high priority

100%! I've ran into the same issue.

> Docs typo-ish improvement

Great catch! PR opened here: https://github.com/tinacms/tinacms.org/pull/1288
